src/lib.rs

```diff
@@ -124,171 +124,401 @@ use proc_macro::TokenStream;
+	let impl_trait = item_impl
+		.trait_
+		.as_ref()
+		.ok_or_else(|| {
+			syn::Error::new_spanned(item_impl.clone(), "missing implementation trait")
+		})?;
 
-    let mut sealed_path = impl_trait.1.segments.clone();
+	let mut sealed_path = impl_trait.1.segments.clone();
 
-    // since `impl for ...` is not allowed, this path will *always* have at least length 1
-    // thus both `first` and `last` are safe to unwrap
-    let syn::PathSegment { ident, arguments } = sealed_path.pop().unwrap().into_value();
-    let seal = seal_name(ident.unraw());
-    sealed_path.push(parse_quote!( #seal ));
-    sealed_path.push(parse_quote!(Sealed));
+	// since `impl for ...` is not allowed, this path will *always* have at least length 1
+	// thus both `first` and `last` are safe to unwrap
+	let syn::PathSegment { ident, arguments } = sealed_path
+		.pop()
+		.unwrap()
+		.into_value();
+	let seal = seal_name(ident.unraw());
+	sealed_path.push(parse_quote!( #seal ));
+	sealed_path.push(parse_quote!(Sealed));
 
-    let self_type = &item_impl.self_ty;
+	let self_type = &item_impl.self_ty;
 
-    // Only keep the introduced params (no bounds), since
-    // the bounds may break in the `#seal` submodule.
-    let (trait_generics, _, where_clauses) = item_impl.generics.split_for_impl();
+	// Only keep the introduced params (no bounds), since
+	// the bounds may break in the `#seal` submodule.
+	let (trait_generics, _, where_clauses) = item_impl
+		.generics
+		.split_for_impl();
 
-    Ok(quote! {
-        #[automatically_derived]
-        impl #trait_generics #sealed_path #arguments for #self_type #where_clauses {}
-        #item_impl
-    })
+	for item in &mut item_impl.items {
+		if let syn::ImplItem::Fn(item) = item {
+			let input = parse_function_arguments(&item.attrs)?;
+
+			let Some((attr, input)) = input else {
+				continue;
+			};
+
+			let attr = attr.clone();
+
+			item.attrs
+				.retain(|x| x != &attr);
+
+			parse_function_implementation(&ident, input, item)?;
+		}
+	}
+
+	Ok(quote! {
+		#[automatically_derived]
+		impl #trait_generics #sealed_path #arguments for #self_type #where_clauses {}
+		#item_impl
+	})
 }
 
 /// Constructs [`syn::Ident`] of a sealing module name.
 fn seal_name<D: fmt::Display>(seal: D) -> syn::Ident {
-    format_ident!("__seal_{}", &seal.to_string().to_snake_case())
+	format_ident!(
+		"__seal_{}",
+		&seal
+			.to_string()
+			.to_snake_case()
+	)
+}
+
+/// Constructs [`syn::Ident`] of a function wrapped by another function
+fn seal_function_name<D: fmt::Display>(seal: D) -> syn::Ident {
+	format_ident!("_{}", &seal.to_string())
+}
+
+fn parse_function_arguments(
+	attrs: &[syn::Attribute],
+) -> Result<Option<(&syn::Attribute, FunctionArguments)>, syn::Error> {
+	for attr in attrs {
+		if attr.path().is_ident("seal") {
+			let value = if let syn::Meta::List(list) = &attr.meta {
+				syn::parse(list.tokens.clone().into())?
+			} else {
+				FunctionArguments::default()
```

Comment:
Replace this with an error.

Let's force the users to write `#[seal(callable)]` or `#[seal(uncallable)]` so the attribute becomes more self-documenting